### ISSUE TITLE
Cannot compile to C++17 standard without -fpermissive

### ISSUE THREAD

**Issue by colin-broderick** [nanodbc/nanodbc#341]
This is related to a conversation in another issue thread. I have managed to narrow down the cause of the issue.

## Environment

* nanodbc version: [34d2ce3](https://github.com/nanodbc/nanodbc/commit/34d2ce31d3b63659ff210bf0d368ccd962e8c3b4)
* DBMS name/version: unixodbc
* ODBC connection string: N/A
* OS and Compiler: Ubuntu 20.04, g++ 9.4.0
* CMake settings: N/A

## Actual behavior

nanodbc cannot compile without -fpersmissive flag if -std=c++17 is set.

## Expected  behavior

Error-free compilation.

## Minimal Working Example

In the nanodbc subfolder, the following command:
```
g++ -c nanodbc.cpp -I .. -lodbc -std=c++17
```
It is specifically the C++17 part that causes the issue to appear. If you choose not to support 17 that's fine of course, but I think a fix would be good so that it can be safely included in other projects with differing standards.

I'm afraid I can't offer a full fix at this time because I don't fully understand the issue. But the issue can be worked around by either
* passing `-fpermissive`, turning the error into a warning, or
* Removing `#define NANODBC_SUPPORT_STRING_VIEW` or adding `#undef NANODBC_SUPPORT_STRING_VIEW` immediately after its definition in nanodbc.h, line 154. This presumably means string view is then not usable. For me that's preferable to compiling non-compliant code as I typically just use a string ref anyway.

## g++ output
```
nanodbc.cpp:5169:82: error: duplicate explicit instantiation of ‘void nanodbc::statement::bind_strings(short int, const T*, std::size_t, std::size_t, nanodbc::statement::param_direction) [with T = char; <template-parameter-1-2> = void; std::size_t = long unsigned int]’ [-fpermissive]
 5169 |         short, const type::value_type*, std::size_t, std::size_t, param_direction);                \
      |                                                                                  ^
nanodbc.cpp:5201:1: note: in expansion of macro ‘NANODBC_INSTANTIATE_BIND_STRINGS’
 5201 | NANODBC_INSTANTIATE_BIND_STRINGS(std::string_view);
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
nanodbc.cpp:5176:24: error: duplicate explicit instantiation of ‘void nanodbc::statement::bind_strings(short int, const T*, std::size_t, std::size_t, const T*, nanodbc::statement::param_direction) [with T = char; <template-parameter-1-2> = void; std::size_t = long unsigned int]’ [-fpermissive]
 5176 |         param_direction);                                                                          \
      |                        ^
nanodbc.cpp:5201:1: note: in expansion of macro ‘NANODBC_INSTANTIATE_BIND_STRINGS’
 5201 | NANODBC_INSTANTIATE_BIND_STRINGS(std::string_view);
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
nanodbc.cpp:5178:95: error: duplicate explicit instantiation of ‘void nanodbc::statement::bind_strings(short int, const T*, std::size_t, std::size_t, const bool*, nanodbc::statement::param_direction) [with T = char; <template-parameter-1-2> = void; std::size_t = long unsigned int]’ [-fpermissive]
 5178 |         short, type::value_type const*, std::size_t, std::size_t, bool const*, param_direction)
      |                                                                                               ^
nanodbc.cpp:5201:1: note: in expansion of macro ‘NANODBC_INSTANTIATE_BIND_STRINGS’
 5201 | NANODBC_INSTANTIATE_BIND_STRINGS(std::string_view);
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
nanodbc.cpp:5169:82: error: duplicate explicit instantiation of ‘void nanodbc::statement::bind_strings(short int, const T*, std::size_t, std::size_t, nanodbc::statement::param_direction) [with T = char16_t; <template-parameter-1-2> = void; std::size_t = long unsigned int]’ [-fpermissive]
 5169 |         short, const type::value_type*, std::size_t, std::size_t, param_direction);                \
      |                                                                                  ^
nanodbc.cpp:5202:1: note: in expansion of macro ‘NANODBC_INSTANTIATE_BIND_STRINGS’
 5202 | NANODBC_INSTANTIATE_BIND_STRINGS(wide_string_view);
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
nanodbc.cpp:5176:24: error: duplicate explicit instantiation of ‘void nanodbc::statement::bind_strings(short int, const T*, std::size_t, std::size_t, const T*, nanodbc::statement::param_direction) [with T = char16_t; <template-parameter-1-2> = void; std::size_t = long unsigned int]’ [-fpermissive]
 5176 |         param_direction);                                                                          \
      |                        ^
nanodbc.cpp:5202:1: note: in expansion of macro ‘NANODBC_INSTANTIATE_BIND_STRINGS’
 5202 | NANODBC_INSTANTIATE_BIND_STRINGS(wide_string_view);
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
nanodbc.cpp:5178:95: error: duplicate explicit instantiation of ‘void nanodbc::statement::bind_strings(short int, const T*, std::size_t, std::size_t, const bool*, nanodbc::statement::param_direction) [with T = char16_t; <template-parameter-1-2> = void; std::size_t = long unsigned int]’ [-fpermissive]
 5178 |         short, type::value_type const*, std::size_t, std::size_t, bool const*, param_direction)
      |                                                                                               ^
nanodbc.cpp:5202:1: note: in expansion of macro ‘NANODBC_INSTANTIATE_BIND_STRINGS’
 5202 | NANODBC_INSTANTIATE_BIND_STRINGS(wide_string_view);
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
nanodbc.cpp:5405:65: error: duplicate explicit instantiation of ‘void nanodbc::table_valued_parameter::bind_strings(short int, const T*, std::size_t, std::size_t) [with T = char; <template-parameter-1-2> = void; std::size_t = long unsigned int]’ [-fpermissive]
 5405 |         short, const type::value_type*, std::size_t, std::size_t);                                 \
      |                                                                 ^
nanodbc.cpp:5432:1: note: in expansion of macro ‘NANODBC_INSTANTIATE_TVP_BIND_STRINGS’
 5432 | NANODBC_INSTANTIATE_TVP_BIND_STRINGS(std::string_view);
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
nanodbc.cpp:5407:90: error: duplicate explicit instantiation of ‘void nanodbc::table_valued_parameter::bind_strings(short int, const T*, std::size_t, std::size_t, const T*) [with T = char; <template-parameter-1-2> = void; std::size_t = long unsigned int]’ [-fpermissive]
 5407 |         short, type::value_type const*, std::size_t, std::size_t, type::value_type const*);        \
      |                                                                                          ^
nanodbc.cpp:5432:1: note: in expansion of macro ‘NANODBC_INSTANTIATE_TVP_BIND_STRINGS’
 5432 | NANODBC_INSTANTIATE_TVP_BIND_STRINGS(std::string_view);
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
nanodbc.cpp:5409:78: error: duplicate explicit instantiation of ‘void nanodbc::table_valued_parameter::bind_strings(short int, const T*, std::size_t, std::size_t, const bool*) [with T = char; <template-parameter-1-2> = void; std::size_t = long unsigned int]’ [-fpermissive]
 5409 |         short, type::value_type const*, std::size_t, std::size_t, bool const*)
      |                                                                              ^
nanodbc.cpp:5432:1: note: in expansion of macro ‘NANODBC_INSTANTIATE_TVP_BIND_STRINGS’
 5432 | NANODBC_INSTANTIATE_TVP_BIND_STRINGS(std::string_view);
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
nanodbc.cpp:5405:65: error: duplicate explicit instantiation of ‘void nanodbc::table_valued_parameter::bind_strings(short int, const T*, std::size_t, std::size_t) [with T = char16_t; <template-parameter-1-2> = void; std::size_t = long unsigned int]’ [-fpermissive]
 5405 |         short, const type::value_type*, std::size_t, std::size_t);                                 \
      |                                                                 ^
nanodbc.cpp:5433:1: note: in expansion of macro ‘NANODBC_INSTANTIATE_TVP_BIND_STRINGS’
 5433 | NANODBC_INSTANTIATE_TVP_BIND_STRINGS(wide_string_view);
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
nanodbc.cpp:5407:90: error: duplicate explicit instantiation of ‘void nanodbc::table_valued_parameter::bind_strings(short int, const T*, std::size_t, std::size_t, const T*) [with T = char16_t; <template-parameter-1-2> = void; std::size_t = long unsigned int]’ [-fpermissive]
 5407 |         short, type::value_type const*, std::size_t, std::size_t, type::value_type const*);        \
      |                                                                                          ^
nanodbc.cpp:5433:1: note: in expansion of macro ‘NANODBC_INSTANTIATE_TVP_BIND_STRINGS’
 5433 | NANODBC_INSTANTIATE_TVP_BIND_STRINGS(wide_string_view);
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
nanodbc.cpp:5409:78: error: duplicate explicit instantiation of ‘void nanodbc::table_valued_parameter::bind_strings(short int, const T*, std::size_t, std::size_t, const bool*) [with T = char16_t; <template-parameter-1-2> = void; std::size_t = long unsigned int]’ [-fpermissive]
 5409 |         short, type::value_type const*, std::size_t, std::size_t, bool const*)
      |                                                                              ^
nanodbc.cpp:5433:1: note: in expansion of macro ‘NANODBC_INSTANTIATE_TVP_BIND_STRINGS’
 5433 | NANODBC_INSTANTIATE_TVP_BIND_STRINGS(wide_string_view);
```

**Comment by mloskot** [comment]
Thank you. I will look into that soon.
We definitely do not want to limit the nanodbc usage to C++14 or any particular C++ language version.

**Comment by colin-broderick** [comment]
Great, thanks for the response.

**Comment by cbroderickgreeve** [comment]
This can be closed as it was seemingly fixed in the PR #353. I have tested and it does now compile without `-fpermissive`.

**Comment by mloskot** [comment]
@cbroderickgreeve Thank you for checking and the confirmation
